ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink

```diff
@@ -156,7 +213,64 @@ if [ "$ret" -eq 1 ]; then
 else
 	PROTO=https
 fi
-force_wan_connection wget -q  -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$PROTO://$(uci get wireguard.mesh_vpn.broker)"
+
+# Remove API path suffix if still present in config
+WGKEX_BROKER_BASE_PATH="$(get_site_string mesh_vpn.wireguard.broker | sed 's|/api/v1/wg/key/exchange||')"
+
+if is_loadbalancing_enabled; then
+	# Use /api/v2, get gateway peer details from broker response
+	WGKEX_BROKER="$PROTO://$WGKEX_BROKER_BASE_PATH/api/v2/wg/key/exchange"
+	logger -p info -t checkuplink "Loadbalancing enabled. Contacting wgkex broker $WGKEX_BROKER"
+	if ! WGKEX_DATA=$(force_wan_connection wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER"); then
+		logger -p err -t checkuplink "Contacting wgkex broker failed, response: $WGKEX_DATA"
+		exit 1
+	fi
+
+	logger -p info -t checkuplink "Got data from wgkex broker: $WGKEX_DATA"
+
+	# Parse the returned JSON in a Lua script, returning the endpoint address, port, pubkey and first allowed IP, separated by newlines
+	if ! data=$(lua /lib/gluon/gluon-mesh-wireguard-vxlan/parse-wgkex-response.lua "$WGKEX_DATA"); then
+		logger -p err -t checkuplink "Parsing wgkex broker data failed"
+		exit 1
+	fi
+
+	logger -p debug -t checkuplink "Successfully parsed wgkex broker data"
+
+	PEER_ADDRESS="$(echo "$data" | sed -n 1p)"
+	PEER_PORT="$(echo "$data" | sed -n 2p)"
+	PEER_PUBLICKEY="$(echo "$data" | sed -n 3p)"
+	PEER_LINKADDRESS=$(echo "$data" | sed -n 4p)
+
+	PEER_ADDRESS="$(resolve_host "$PEER_ADDRESS")"
+	PEER_ENDPOINT="$(combine_ip_port "$PEER_ADDRESS" "$PEER_PORT")"
+
+else
+	# Use /api/v1, get gateway peer details from config
+	WGKEX_BROKER="$PROTO://$WGKEX_BROKER_BASE_PATH/api/v1/wg/key/exchange"
+	logger -p info -t checkuplink "Loadbalancing disabled. Contacting wgkex broker $WGKEX_BROKER"
+	if ! force_wan_connection wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER"; then
+		logger -p err -t checkuplink "Contacting wgkex broker failed"
+		exit 1
+	fi
+
+	# Get the number of configured peers and randomly select one
+	NUMBER_OF_PEERS=$(uci -q show wireguard | grep -E -ce "peer_[0-9]+.endpoint")
+	PEER="$(awk -v min=1 -v max="$NUMBER_OF_PEERS" 'BEGIN{srand(); print int(min+rand()*(max-min+1))}')"
```

Comment:
I think we should still change this for v1 because I recently found out that this random is "random" based on the system time, maybe there is another option to determine the random without depending on the system time. Which is why so many clients choose the same gateway from the list but this shouldn't be a blocker for a merge as the v2 offers significantly better advantages for load balancing

Reply:
It seems that our awk implementation for `srand` only uses second-precision, that is indeed not very helpful:
```sh
# i=0; while [[ $i -lt 15 ]]; do date -Ins | tr -d '\n' && echo -n " rand()=" && awk 'BEGIN{srand(); print rand()}'; i=$((i+1)); done
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
2024-03-26T13:32:18,000000000+01:00 rand()=0.764051
```
I can confirm this is the same with Ubuntu 22.04 on a desktop system (but it does show nanoseconds on `date --iso=ns`).

As `$RANDOM` is not available in busybox sh here, an alternative would be to use urandom + hexdump: 
```sh
# i=0; while [[ $i -lt 15 ]]; do i=$((i+1)); date -Ins | tr -d '\n' && echo -n " rand()=" && hexdump -n 4 -e '"%u"' </dev/urandom && echo; done
2024-03-26T13:36:00,000000000+01:00 rand()=2986735940
2024-03-26T13:36:00,000000000+01:00 rand()=234726918
2024-03-26T13:36:00,000000000+01:00 rand()=4249724515
2024-03-26T13:36:00,000000000+01:00 rand()=165910888
2024-03-26T13:36:00,000000000+01:00 rand()=2380433412
2024-03-26T13:36:00,000000000+01:00 rand()=3772558562
2024-03-26T13:36:00,000000000+01:00 rand()=277626919
2024-03-26T13:36:00,000000000+01:00 rand()=2142351523
2024-03-26T13:36:00,000000000+01:00 rand()=1096050454
2024-03-26T13:36:00,000000000+01:00 rand()=1464321489
2024-03-26T13:36:00,000000000+01:00 rand()=2153772455
2024-03-26T13:36:00,000000000+01:00 rand()=2246216180
2024-03-26T13:36:00,000000000+01:00 rand()=3724409659
2024-03-26T13:36:00,000000000+01:00 rand()=2829435099
2024-03-26T13:36:00,000000000+01:00 rand()=4106529008
```

Reply:
I opened a separate PR for this: https://github.com/freifunk-gluon/community-packages/pull/101